File: leetcode/Algorithms/300.LongestIncreasingSubsequence/Solution.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dp, result = [1] * len(nums), 0
        for i in xrange(len(nums)):
            for j in xrange(i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], dp[j] + 1)
            result = max(dp[i], result)
        return result

File: leetcode/Algorithms/300.LongestIncreasingSubsequence/Solution.java
import java.util.Arrays;

public class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] dp = new int[nums.length];
        int result = 0;
        for (int num : nums) {
            int position = Arrays.binarySearch(dp, 0, result, num);
            if (position < 0) {
                position = -(position + 1);
                dp[position] = num;
            }
            if (position == result) result++;
        }
        return result;
    }
}
